rex-engine/src/codec.rs

```diff
@@ -472,6 +505,49 @@ where
     }
 }
 
+impl<T, E> Decode for Result<T, E>
+where
+    T: Decode,
+    E: Decode,
+{
+    fn try_decode(v: &Expr) -> Result<Self, Error> {
+        match v {
+            Expr::Named(_id, _span, name, x) if name == "Ok" => {
+                Ok(Ok(T::try_decode(x)?))
+            }
+            Expr::Named(_id, _span, name, x) if name == "Err" => {
+                Ok(Err(E::try_decode(x)?))
+            }
+            _ => Err(Error::ExpectedTypeGotValue {
+                expected: Type::Result(
+                    Box::new(Type::Var(Id::new())),
+                    Box::new(Type::Var(Id::new()))),
+                got: v.clone(),
+            }),
+        }
+    }
+}
+
+impl<T> Decode for Option<T>
+where
+    T: Decode
+{
+    fn try_decode(v: &Expr) -> Result<Self, Error> {
+        match v {
+            Expr::Named(_id, _span, name, x) if name == "Some" => {
+                Ok(Some(T::try_decode(x)?))
+            }
+            Expr::Named(_id, _span, name, _x) if name == "None" => {
+                Ok(None)
+            }
+            _ => Err(Error::ExpectedTypeGotValue {
+                expected: Type::Option(Box::new(Type::Var(Id::new()))),
```

Comment:
same comment as before, require `T: ToType` and avoid the unhelpful `Type::Var(..)`